src/test/kotlin/com/terraformation/backend/documentproducer/db/VariableValueStoreTest.kt

```diff
@@ -564,6 +570,106 @@ class VariableValueStoreTest : DatabaseTest(), RunsAsUser {
+
+      store.upgradeSectionValueVariables(mapOf(oldVariableId to newVariableId))
+
+      val newValues =
+          store.listValues(projectId = projectId, variableIds = listOf(sectionVariableId))
+      val newTextSectionValue = newValues.single { it.listPosition == 0 }
+      val newReferringSectionValue = newValues.single { it.listPosition == 1 }
+
+      assertEquals(
+          oldTextSectionValue, newTextSectionValue, "Text value should not have been modified")
+      assertNotEquals(
+          oldReferringSectionValueId,
+          newReferringSectionValue.id,
+          "Value with variable reference should have been replaced")
+      assertEquals(
+          SectionValueVariable(
+              newVariableId, VariableUsageType.Injection, VariableInjectionDisplayStyle.Block),
+          newReferringSectionValue.value,
+          "Section should refer to new variable")
+    }
+
+    @Test
+    fun `does not modify sections with up-to-date variable references`() {
+      val projectId = inserted.projectId
+      insertDocumentTemplate()
+      insertVariableManifest()
+      insertDocument()
+      val sectionVariableId = insertVariableManifestEntry(insertSectionVariable())
+
+      val oldVariableId = insertTextVariable()
+
+      // The first version of the section references the outdated variable, there is no project
+      // value for this variable yet
+      insertSectionValue(listPosition = 0, textValue = "some text", variableId = sectionVariableId)
+      insertSectionValue(
+          displayStyle = VariableInjectionDisplayStyle.Block,
+          listPosition = 1,
+          usageType = VariableUsageType.Injection,
+          usedVariableId = oldVariableId,
+          variableId = sectionVariableId)
+
+      // The variable is upgraded at some point
+      val newVariableId =
+          insertTextVariable(
+              insertVariable(replacesVariableId = oldVariableId, type = VariableType.Text))
+
+      // The section has already been updated to use the new variable ID
+      insertSectionValue(
+          listPosition = 0, textValue = "some updated text", variableId = sectionVariableId)
+      insertSectionValue(
+          displayStyle = VariableInjectionDisplayStyle.Block,
+          listPosition = 1,
+          usageType = VariableUsageType.Injection,
+          usedVariableId = newVariableId,
+          variableId = sectionVariableId)
+
+      // A project value is added to the new variable
+      insertValue(textValue = "referenced text", variableId = newVariableId)
+
+      every { user.canUpdateInternalVariableWorkflowDetails(projectId) } returns true
+
+      store.upgradeSectionValueVariables(mapOf(oldVariableId to newVariableId))
+
+      // There should not be any values updated since the up-to-date version of the section value
+      // is already referencing the upgraded variable
+      verify(exactly = 1) { store.updateValues(emptyList(), false) }
```

Comment:
This is testing an implementation detail; if we add code to skip the `updateValues` call if the list is empty, this test will break even though the end result of the upgrade method didn't change at all.

A better approach, I think, would be to fetch the value of the section variable before and after the upgrade call and assert that the two are equal, since the behavior we want in this scenario is that the value was left as-is.

Reply:
Makes sense!